File: py/behavioral/rare_memento.py
# Memento captures state in a separate object so it can be restored later.
# Modern Python note: snapshots are useful, but dataclasses, copy, persistence, or
# event logs are often clearer than a dedicated Memento type.
class EditorMemento:
    def __init__(self, text):
        self.text = text


class Editor:
    def __init__(self):
        self.text = ""

    def type(self, words):
        self.text += words

    def save(self):
        return EditorMemento(self.text)

    def restore(self, memento):
        self.text = memento.text


if __name__ == "__main__":
    editor = Editor()
    editor.type("draft")
    saved = editor.save()
    editor.type(" with edits")
    editor.restore(saved)
    print(editor.text)

File: csharp/behavioral/rare_memento.cs
namespace Gof.ModernCSharp.Behavioral;

// Memento captures state in a separate object so it can be restored later.
// Modern C# note: records, snapshots, persistence, or event sourcing may be
// clearer than a dedicated Memento type.
public static class RareMemento
{
    sealed record EditorMemento(string Text);
    sealed class Editor
    {
        public string Text { get; private set; } = "";
        public void Type(string words) => Text += words;
        public EditorMemento Save() => new(Text);
        public void Restore(EditorMemento memento) => Text = memento.Text;
    }

    public static string Run()
    {
        var editor = new Editor();
        editor.Type("draft");
        var saved = editor.Save();
        editor.Type(" with edits");
        editor.Restore(saved);
        return editor.Text;
    }
}
